modules/xcm-bridge-hub-router/src/lib.rs

```diff
@@ -0,0 +1,557 @@
+/// Maximal size of the XCM message that may be sent over bridge.
+///
+/// This should be less than the maximal size, allowed by the messages pallet, because
+/// the message itself is wrapped in other structs and is double encoded.
+pub const HARD_MESSAGE_SIZE_LIMIT: u32 = 32 * 1024;
+
+/// The target that will be used when publishing logs related to this pallet.
+///
+/// This doesn't match the pattern used by other bridge pallets (`runtime::bridge-*`). But this
+/// pallet has significant differences with those pallets. The main one is that is intended to
+/// be deployed at sending chains. Other bridge pallets are likely to be deployed at the separate
+/// bridge hub parachain.
+pub const LOG_TARGET: &str = "xcm::bridge-hub-router";
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::config]
+	pub trait Config<I: 'static = ()>: frame_system::Config {
+		/// Benchmarks results from runtime we're plugged into.
+		type WeightInfo: WeightInfo;
+
+		/// Universal location of this runtime.
+		type UniversalLocation: Get<InteriorMultiLocation>;
+		/// The bridged network that this config is for if specified.
+		/// Also used for filtering `Bridges` by `BridgedNetworkId`.
+		/// If not specified, allows all networks pass through.
+		type BridgedNetworkId: Get<Option<NetworkId>>;
+		/// Configuration for supported **bridged networks/locations** with **bridge location** and
+		/// **possible fee**. Allows to externalize better control over allowed **bridged
+		/// networks/locations**.
+		type Bridges: ExporterFor;
+
+		/// Origin of the sibling bridge hub that is allowed to report bridge status.
+		type BridgeHubOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+		/// Actual message sender (`HRMP` or `DMP`) to the sibling bridge hub location.
+		type ToBridgeHubSender: SendXcm;
+		/// Underlying channel with the sibling bridge hub. It must match the channel, used
+		/// by the `Self::ToBridgeHubSender`.
+		type WithBridgeHubChannel: XcmChannelStatusProvider;
+
+		/// Additional fee that is paid for every byte of the outbound message.
+		type ByteFee: Get<u128>;
+		/// Asset that is used to paid bridge fee.
+		type FeeAsset: Get<AssetId>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+	#[pallet::hooks]
+	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {
+		fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+			// TODO: make sure that `WithBridgeHubChannel::is_congested` returns true if either
+			// of XCM channels (outbound/inbound) is suspended. Because if outbound is suspended
+			// that is definitely congestion. If inbound is suspended, then we are not able to
+			// receive the "report_bridge_status" signal (that maybe sent by the bridge hub).
+
+			// if the channel with sibling/child bridge hub is suspended, we don't change
+			// anything
+			if T::WithBridgeHubChannel::is_congested() {
```

Comment:
The base fee then need to include both bridge fee and HRMP fee. And then instead of the `BridgeFee * BridgeFeeFactor + HrmpFee * HrmpFeeFactor`, we'll temporary have the `(BridgeFee  + HrmpFee) * BridgeFeeFactor` until https://github.com/paritytech/cumulus/pull/2342 is implemented. This is fine, because the exponential factor is present in the formula.